### PR TITLE
Add test for transaction has too high prioritization fee

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem
Web3.js has a test started to fail against tip of validator master. The test sends a transaction requesting 2sol as prioritization fee, with payer account has 2sol balance. Test correctly expects the transaction to failed due to `Err(TransactionError::InsufficientFundsForFee)` therefore not landed.

#### Summary of Changes
- Add tests to confirm the transaction base fee calculation and payer account validation behave correctly.

Fixes #

### CODE REVIEW

**Review by apfitzge — COMMENTED** [review]
tests look good, some small comments and questions about values

File: runtime/src/accounts.rs [hunk 1]
```diff
@@ -4257,4 +4257,67 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_load_accounts_too_high_prioritization_fee() {
+        solana_logger::setup();
+        let lamports_per_signature = 5000;
+        let request_units = 1_000_000;
+        let request_unit_price = 2_000_000_000; // 2B micro-lamports/CU
+                                                // prioritization_fee = 2B micro-lamports/CU * 1M CUs = (2B / 1M) * 1M = 2B lamports
+        let prioritization_fee = 2_000_000_000;
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply runtime/src/accounts.rs:4268]: `cargo fmt --all` pushed comment at ln 4267 all the way to right, made the intention less obvious. But in any case, utilizing `MICRO_LAMPORTS_PER_LAMPORT` is better than all the static numbers, even better to just call `let prioritization_fee = prioritization_fee_details.get_fee();` 😄

File: runtime/src/accounts.rs [hunk 2]
```diff
@@ -4257,4 +4257,67 @@ mod tests {
...
+        let mut error_counters = TransactionErrorMetrics::default();
+        let keypair = Keypair::new();
+        let key0 = keypair.pubkey();
+
+        // set up account with 2B lamports
+        let account = AccountSharedData::new(prioritization_fee, 0, &Pubkey::default());
+        accounts.push((key0, account));
+
+        let instructions = &[
+            solana_sdk::compute_budget::ComputeBudgetInstruction::set_compute_unit_limit(
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply runtime/src/accounts.rs:4280]: good call.

File: runtime/src/accounts.rs [hunk 3]
```diff
@@ -4257,4 +4257,67 @@ mod tests {
...
+        let prioritization_fee = 2_000_000_000;
+
+        let mut accounts: Vec<TransactionAccount> = Vec::new();
+        let mut error_counters = TransactionErrorMetrics::default();
+        let keypair = Keypair::new();
+        let key0 = keypair.pubkey();
+
+        // set up account with 2B lamports
+        let account = AccountSharedData::new(prioritization_fee, 0, &Pubkey::default());
+        accounts.push((key0, account));
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply runtime/src/accounts.rs:4277]: ✅

File: runtime/src/accounts.rs [hunk 4]
```diff
@@ -4257,4 +4257,67 @@ mod tests {
...
+            true,
+            true,
+            true,
+            false,
+        );
+        assert_eq!(fee, lamports_per_signature + prioritization_fee);
+
+        // assert fail to load account with 2B lamport balance for transaction asking for 2B
+        // lamports as prioritization fee.
+        let loaded_accounts = load_accounts_with_fee(
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply runtime/src/accounts.rs:4308]: ✅

File: runtime/src/accounts.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply runtime/src/accounts.rs:4268]: It doesn't really matter if it is 2B lamports, I just try to recreate same condition as web3.js test in question. (https://github.com/solana-labs/solana-web3.js/pull/1238/files), it funded account with 2Sol and requested 2sol as priority fee.

File: runtime/src/accounts.rs [hunk 5]
```diff
@@ -4257,4 +4257,67 @@ mod tests {
...
+            lamports_per_signature,
+            &mut error_counters,
+            None,
+        );
+
+        assert_eq!(error_counters.insufficient_funds, 1);
+        assert_eq!(loaded_accounts.len(), 1);
+        assert_eq!(
+            loaded_accounts[0].clone(),
+            (Err(TransactionError::InsufficientFundsForFee), None,),
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply runtime/src/accounts.rs:4320]: until `cargo fmt` does its job, will do manual fix 😸

File: runtime/src/bank/tests.rs [hunk 6]
```diff
@@ -10892,6 +10892,45 @@ fn test_calculate_fee_compute_units() {
...
+
+    let message = SanitizedMessage::try_from(Message::new(
+        &[
+            ComputeBudgetInstruction::set_compute_unit_limit(request_units),
+            ComputeBudgetInstruction::set_compute_unit_price(request_unit_price),
+        ],
+        Some(&Pubkey::new_unique()),
+    ))
+    .unwrap();
+    for support_set_accounts_data_size_limit_ix in [true, false] {
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply runtime/src/bank/tests.rs:10915]: Ah copy/paste .... thanks for pointing it out, changed to simply use `true` to be consistent

File: runtime/src/accounts.rs [hunk 7]
```diff
@@ -4257,4 +4261,64 @@ mod tests {
...
+
+    #[test]
+    fn test_load_accounts_too_high_prioritization_fee() {
+        solana_logger::setup();
+        let lamports_per_signature = 5000_u64;
+        let request_units = 1_000_000_u32;
+        let request_unit_price = 2_000_000_000_u64;
+        let prioritization_fee_details = PrioritizationFeeDetails::new(
+            PrioritizationFeeType::ComputeUnitPrice(request_unit_price),
+            request_units as u64,
```
Comment (apfitzge) [inline runtime/src/accounts.rs:4273]: unrelated to the scope of this PR - it seems odd that this takes `request_units` as a `u64`, but `set_compute_unit_limit` takes it as a `u32`

File: runtime/src/bank/tests.rs [hunk 8]
```diff
@@ -10892,6 +10892,44 @@ fn test_calculate_fee_compute_units() {
...
+fn test_calculate_prioritization_fee() {
+    let fee_structure = FeeStructure {
+        lamports_per_signature: 1,
+        ..FeeStructure::default()
+    };
+
+    let request_units = 1_000_000;
+    let request_unit_price = 2_000_000_000; // 2B micro-lamports/CU
+                                            // prioritization_fee = 2B micro-lamports/CU * 1M CUs = (2B / 1M) * 1M = 2B lamports
+    let prioritization_fee = 2_000_000_000;
```
Comment (apfitzge) [inline runtime/src/bank/tests.rs:10905]: sorry, could we also get this one changed like the test in accounts.rs?

File: runtime/src/accounts.rs (hunk 7 above)
Comment (taozhu-chicago) [reply runtime/src/accounts.rs:4273]: yea, there is inconsistency between https://github.com/solana-labs/solana/blob/master/sdk/src/compute_budget.rs#L40 and https://github.com/solana-labs/solana/blob/master/program-runtime/src/compute_budget.rs#L34

**Pull request merged**